src/Workspaces/Core/Portable/Storage/SQLite/SQLitePersistentStorage_Threading.cs

```diff
@@ -0,0 +1,102 @@
+        /// deadlocks.  Specifically, consider two threads doing the following:
+        ///
+        /// Thread A starts a transaction that starts as a reader, and later attempts to perform a
+        /// write. Thread B is a writer (either started that way, or started as a reader and
+        /// promoted to a writer first). B holds a RESERVED lock, waiting for readers to clear so it
+        /// can start writing. A holds a SHARED lock (it's a reader) and tries to acquire RESERVED
+        /// lock (so it can start writing).  The only way to make progress in this situation is for
+        /// one of the transactions to roll back. No amount of waiting will help, so when SQLite
+        /// detects this situation, it doesn't honor the busy timeout.
+        ///
+        /// To prevent this scenario, we control our access to the db explicitly with operations that
+        /// can concurrently read, and operations that exclusively write.
+        /// 
+        /// All code that reads or writes from the db should go through this.
+        /// </summary>
+        private readonly ConcurrentExclusiveSchedulerPair _readerWriterLock = new ConcurrentExclusiveSchedulerPair();
+
+        // Inner class used so that we can generically create and pool StrongBox<TArg> instances. We
+        // use those to provide an allocation-free means of calling
+        // <c>Task.Factory.StartNew(func<object>, object, ...)</c>.
+        private static class Threading<TArg, TResult>
+            where TArg : struct
+        {
+            private static readonly Stack<StrongBox<(Func<TArg, TResult> func, TArg arg)>> s_boxes =
+                 new Stack<StrongBox<(Func<TArg, TResult> func, TArg arg)>>();
+
+            private static StrongBox<(Func<TArg, TResult> func, TArg arg)> GetBox()
+            {
+                lock (s_boxes)
+                {
+                    if (s_boxes.Count > 0)
+                    {
+                        return s_boxes.Pop();
+                    }
+
+                    return new StrongBox<(Func<TArg, TResult> func, TArg arg)>();
+                }
+            }
+
+            private static void ReturnBox(StrongBox<(Func<TArg, TResult> func, TArg arg)> box)
+            {
+                lock (s_boxes)
+                {
+                    s_boxes.Push(box);
+                }
+            }
+
+            [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/36114", AllowCaptures = false)]
+            public static Task<TResult> PerformTask(Func<TArg, TResult> func, TArg arg, TaskScheduler scheduler, CancellationToken cancellationToken)
+            {
+                // Use a pooled strongbox so we don't box the provided arg.
+                var box = GetBox();
+                box.Value = (func, arg);
+
+                return Task.Factory.StartNew(Callback, box, cancellationToken, TaskCreationOptions.None, scheduler);
```

Comment:
As you care about allocations, you presumably want to avoid the delegate allocation that's going to result from passing Callback as a method group.

Reply:
Yes.  that was an outdated commit.  It's already been moved (along with all other lambdas) to a static field. :)